[4.0] WebAsset: redesign

Fedik opened the pull request:
### Summary of Changes

As result of discussion in #25309 I have decided to make some changes.
The changes compared to existing WebAsset logic:
 - Now the assets are  managed separately per type: `script`, `style` etc. (possible to add webcomponents also)
 - WebAssetItem represent a single file (instead of Collection of `js`, `css`, as before). This also change JSON schema, and simplify attributes editing (@wilsonge I will do separated pull later)
 - The Collection of `js`, `css` now represents by "empty" WebAssetItem with "cross-dependency" (called 'preset')
 - An "attach" method was removed, now it part of rendering process
 - AssetManager now allow to "register" an asset file as AssetItem, eg `registerScript($asset)`/`registerStyle($asset)`/`registerAsset('foobar', $asset)`

### Questions that need to discuss:

**Asset rendering and backward compatibility.**
~In current implementation, I have put all assets to _scripts/_stylesheet property of document (before rendering).
This allow to have all scripts/styles `onBeforeCompileHead` , but this also has a drawback.
It not allow to use AssetManager while this event (see #24858 and #24848).~

~So I think maybe we can break here a bit?
I mean, do not merge Asset files with _scripts/_stylesheet, but keep them in AssetManager.
Potential b.c. here that _scripts/_stylesheet will not contain files from  active Asset at `onBeforeCompileHead`. (it also allow to drop WebAssetBeforeAttachEvent as useless)
Thoughts?~
(I have split them)

 ~**Inline scrips/styles**
I have skipped it for now, this part require full drop of _script/_style and move to AssetManager.
But because now an assets are managed separately (per type) Inline scrips/styles should be more easy to implement it in future.
In theory it is possible just ignore these properties, and proxy add[Script/Style]Declaration, but I am not sure currently how it will affect 3d extensions.
Thoughts?~
(This is for another pull)

**Other questions:**
~Make webcomponents to use AssetManager (no b.c.) Yes/No?~ (see #27268)
~Proxy scriptOptions to AssetManager (minimum b.c.) Yes/No?~ (can be in another pull)

### Testing Instructions
Apply patch run `npm install` and make sure everything still work as before

### Methods comparison
For whom interesting

 AssetManager | WP
---------------------|--------
`$wa->register[Script,Style,Foobar]()` |  `wp_register_[script,style]()`
`$wa->use[Script,Style,Foobar]()` <br> `$wa->registerAndUse[Script,Style,Foobar]()`  | `wp_enqueue_[script,style]()`
`$wa->disable[Script,Style,Foobar]()` | `wp_dequeue_[script,style]()`
n.a. | `wp_add_inline_[script,style]()`

### Documentation Changes Required

yeap

@wilsonge @mbabker please review

#### For references:
doc by @mbabker [Joomla Document/WebAsset Analysis](https://docs.google.com/document/d/1ajboeOkOYq9Q9OdSH0d8fSa4lv6qpiDodlv0igSGZlk)
previous pulls #23463 #22435

Fedik committed:
WebAssetManager redesign

Squashed commit of the following:

commit 306b7855e74b069f70fe9e8d5dc9059b3429e46e
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 20:38:36 2019 +0300

    deprecated a bit

commit 59dfcd7385d7c7259946350f1793b7f36c4a6e6d
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 20:16:14 2019 +0300

    webcomponent and webasset

commit f76df9f803e9de971317064df320cb0a297e69ef
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 19:28:02 2019 +0300

    interfaces

commit 9c0d929b943c685e4e772b070a7c649979a88086
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 18:46:57 2019 +0300

    clean up around

commit 16256fcf41f71207b013a5803c077c64e22293e4
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 18:24:17 2019 +0300

    Drop $type property of AssetItem

commit b741f1cf381728e89fdb5085631f3c5fea53b9db
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 17:40:17 2019 +0300

    more replaces of HTMLHelper to registerStyle

commit 21dd619300eab67398164d14d59cddd2e6527c35
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 17:20:43 2019 +0300

    FormValidateAssetItem

commit dfc7727bedc9fa6d77f10532e7cfc7febfdfda9e
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 17:12:24 2019 +0300

    cms/html use registerScript

commit d7f051e47a575b0f98c54dc304b4738630331869
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 16:29:05 2019 +0300

    Check for AttachBehavior, fix cross-dependency, remove Dispatcher

commit b22e05b02577cdbf0e32580babb6683d5c2398f8
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 15:29:06 2019 +0300

    Revert document changes

commit c605ec71c0b70372a6e0c0a02e77815bd8fc5fbd
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 15:22:32 2019 +0300

    Revert scripts/styles renderers to use previous rendering

commit fc3b9a8210df93c3d9ae3489d0dd39e76c66c41f
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 15:16:11 2019 +0300

    Merge assets with an existing, to keep b.c.

commit 91e283c538dcdd2c384b6edc25a3dd3fdb675078
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 14:56:06 2019 +0300

    flexible registerAsset()

commit 6c00787b19a8c77518597089c2ec02b447316e3a
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 14:27:26 2019 +0300

    enable to use, group to preset

commit 62ce690c3856501166686d06136746f5063c1124
Merge: c79d2c0343 511547adb0
Author: Fedik <getthesite@gmail.com>
Date:   Sat Aug 3 12:57:07 2019 +0300

    Sync. Merge branch '4.0-dev' into asset3

commit c79d2c03435c89047e5ae38f3d86770bca9d9914
Author: Fedik <getthesite@gmail.com>
Date:   Thu Aug 1 20:34:40 2019 +0300

    help with debuging, fix of fallback

commit db921a61e83326a35f5f4fffa512bcdd85771b6f
Author: Fedik <getthesite@gmail.com>
Date:   Thu Aug 1 19:50:33 2019 +0300

    couple fixes

commit 3904911f561b03b4cddba5d75c1a1d6efcb15bcf
Author: Fedik <getthesite@gmail.com>
Date:   Tue Jul 30 22:05:10 2019 +0300

    quick proxy for addScript/Style

commit 496baf13bcd23b2ba9dd7032abcdfd0c018180ae
Author: Fedik <getthesite@gmail.com>
Date:   Tue Jul 30 21:42:08 2019 +0300

    make something working

commit 36b730e133806b1d4ae2f068dcf334b15e8dca74
Author: Fedik <getthesite@gmail.com>
Date:   Tue Jul 30 20:50:45 2019 +0300

    update settings.json and provideAssets generation

commit 45871ae45f0e30dc1749e814ed39623e8a4627d8
Author: Fedik <getthesite@gmail.com>
Date:   Tue Jul 30 20:12:18 2019 +0300

    update settings.json

commit e863a2aae94aa03f04801266ddaf7f5bbf6dcd78
Author: Fedik <getthesite@gmail.com>
Date:   Tue Jul 30 19:25:37 2019 +0300

    use style

commit 80f71471b95d8aced88bc7c5a571244e7b12cc5d
Author: Fedik <getthesite@gmail.com>
Date:   Tue Jul 30 18:46:33 2019 +0300

    update asset.json file

commit 18be7d219a8818d28c6c45afd894ebdcafdc208b
Author: Fedik <getthesite@gmail.com>
Date:   Mon Jul 29 21:28:33 2019 +0300

    add lock, update registry file

commit ac6c13a97a52ec828cab2815cfe6429a505f631b
Author: Fedik <getthesite@gmail.com>
Date:   Mon Jul 29 19:59:19 2019 +0300

    fix enableDependencies() per type

commit 4e262801cce19feb3278551e61786d9a993495a9
Author: Fedik <getthesite@gmail.com>
Date:   Mon Jul 29 19:41:01 2019 +0300

    fix sorting

commit 78e102637f3025da077d73e3ce97c093389139a8
Author: Fedik <getthesite@gmail.com>
Date:   Mon Jul 29 19:12:59 2019 +0300

    EenableDependencies per type

commit 8cef0a5ab0555b6317d1953533562654b13ce60f
Author: Fedik <getthesite@gmail.com>
Date:   Sun Jul 28 19:09:36 2019 +0300

    update WebAssetManager

commit 50fdc42a24036464c3729801ce202798965db428
Author: Fedik <getthesite@gmail.com>
Date:   Sun Jul 28 16:18:13 2019 +0300

    update WebAssetItem

commit 897c382170e35ed6cf26d9ee9619637eb4b422ef
Author: Fedik <getthesite@gmail.com>
Date:   Sun Jul 28 14:42:12 2019 +0300

    update schema

Fedik committed:
dog

Fedik committed:
phpcs

Fedik committed:
phpcs

Fedik committed:
Listen to changes in the registry

Fedik committed:
will be changed by listener anyway

Fedik committed:
Listen only changes of active assets

Fedik committed:
now the same but without recursion loop

Fedik committed:
phpcs

Fedik committed:
Add description to an asset item interface

Fedik committed:
Do not put asset items to _scripts/_stylesheet properties of a document

Fedik committed:
correct dependencies

Fedik committed:
remove unused event

C-Lodder commented on administrator/components/com_modules/tmpl/modules/default_batch_body.php:
```diff
@@ -32,7 +32,7 @@
 Text::script('JGLOBAL_SELECT_PRESS_TO_SELECT');
 
-Factory::getDocument()->getWebAssetManager()->enableAsset('choicesjs');
+Factory::getDocument()->getWebAssetManager()->usePreset('choicesjs');
```
Rather than `usePreset`, would it not be better to use something like one of the following:

- `useVendor()`
- `useDependency()`
- `importVendor()`
- `loadVendor()`
- `importDependency()`
- `loadDependency()`

?

Fedik replied:
`usePreset` is a second that I got in my head, first was `useGroup` :smiley: 
well, can be anything, I do not mind, 

`preset` it is a kind of "collection" asset, that do not provide any files, but just enable "depended" assets.

In case of `->usePreset('choicesjs')` this will enable `choicesjs` preset, which has dependecies to `choicesjs` css and choicesjs js assets, so it the same as calling:
`->useStyle('choicesjs')->useScript('choicesjs')`

look:
https://github.com/joomla/joomla-cms/blob/a038be877f8668f3bd9c48e0fee82211b259802f/build/build-modules-js/settings.json#L137-L156

I think, nothing from your suggestion are fit this case :wink:

infograf768 committed:
Merge branch '4.0-dev' into asset-4

infograf768 commented:
This solves the issue I had with specific xx-XX.css admin lang which is now loaded after bootstrap and template css.
```
	<link href="/installmulti/joomla40/media/system/css/fields/switcher.min.css?a1bacb193a9e68d6496252f11a1d20e4" rel="stylesheet" />
	<link href="/installmulti/joomla40/administrator/templates/atum/css/bootstrap.min.css?a1bacb193a9e68d6496252f11a1d20e4" rel="stylesheet" />
	<link href="/installmulti/joomla40/administrator/templates/atum/css/template-rtl.min.css?a1bacb193a9e68d6496252f11a1d20e4" rel="stylesheet" />
	<link href="/installmulti/joomla40/administrator/language/fa-IR/fa-IR.css?a1bacb193a9e68d6496252f11a1d20e4" rel="stylesheet" />
	<link href="/installmulti/joomla40/administrator/templates/atum/css/fontawesome.min.css?a1bacb193a9e68d6496252f11a1d20e4" rel="stylesheet" />
etc.
```

infograf768 commented:
@Fedik 
Shall I set this to tested OK as I just tested the css?

Fedik commented:
@infograf768 yes if other part of the site still works good for you :wink: 
but the whole patch need to be reviewed by @wilsonge and other

Fedik commented:
well, I have to fix a conflict first

wilsonge commented:
I'm planning on looking at this and I'm aware of it - but I'm really busy with gsoc stuff as it ends this week. 20-30th I'm on holiday so I'm unlikely to be able to properly look at it until after I'm back from holiday. So please don't close this - and sorry it's taking me time to review it properly!

Fedik committed:
Sync. Merge branch '4.0-dev' into asset-4

 Conflicts:
	administrator/templates/atum/index.php
	administrator/templates/atum/joomla.asset.json
	layouts/joomla/form/field/accesslevel-fancy-select.php

Fedik committed:
"after conflict" fix

Fedik committed:
template.js to asset

Fedik committed:
Merge branch 'asset-4' of github.com:Fedik/joomla-cms into asset-4

Fedik committed:
Sync. Merge branch '4.0-dev' into asset-4

 Conflicts:
	libraries/src/Event/WebAsset/WebAssetBeforeAttachEvent.php
	libraries/src/WebAsset/WebAssetManager.php
	libraries/src/WebAsset/WebAssetRegistry.php

Fedik committed:
Sync. Merge branch '4.0-dev' into asset-4

Fedik committed:
Sync. Merge branch '4.0-dev' into asset-4

 Conflicts:
	administrator/components/com_modules/tmpl/modules/default_batch_body.php

C-Lodder commented:
I have tested this item :white_check_mark: successfully on 5b2e082f9ac3448787a1d9d2820b3c4b0679713d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25775">issues.joomla.org/tracker/joomla-cms/25775</a>.</sub>

C-Lodder commented:
@Fedik A far better approach for the web assets

@wilsonge Have you managed to look at this? The sooner a decision is made about this, the better

Fedik committed:
Sync. Merge remote-tracking branch 'joomla/4.0-dev' into asset-4

 Conflicts:
	plugins/system/debug/debug.php

Fedik committed:
fix after conflict

Fedik committed:
Sync. Merge branch '4.0-dev' into asset-4

 Conflicts:
	administrator/templates/atum/error_full.php
	administrator/templates/atum/error_login.php
	administrator/templates/atum/index.php
	administrator/templates/atum/joomla.asset.json
	administrator/templates/atum/login.php
	templates/cassiopeia/index.php

Fedik committed:
update atum/component

Fedik committed:
assets to use relative uri to media, instead of full

Quy commented on administrator/templates/atum/component.php:
```diff
@@ -18,17 +18,14 @@
+$wa->usePreset($assetName);
 
 // Load specific language related CSS
-HTMLHelper::_('stylesheet', 'administrator/language/' . $lang->getTag() . '/' . $lang->getTag() . '.css', ['version' => 'auto']);
```
Remove `use Joomla\CMS\HTML\HTMLHelper;`.

Quy commented on templates/cassiopeia/index.php:
```diff
@@ -31,11 +31,13 @@
+$wa->usePreset($assetName);
 
 // Load specific language related CSS
-HTMLHelper::_('stylesheet', 'language/' . $lang->getTag() . '/' . $lang->getTag() . '.css', array('version' => 'auto'));
```
Remove `use Joomla\CMS\HTML\HTMLHelper;`.

Fedik committed:
remove unused HTMLHelper

Fedik committed:
LangActiveAssetItem

alikon commented:
where we are on this @release-joomla  postpone to 5.x ?

Fedik commented:
5.x will be to late, it one of "now or never" :)

alikon commented:
@mbabker , @wilsonge , @HLeithner

Fedik committed:
Improve preset handling

Fedik committed:
Correcting use of assets in template files

Fedik commented:
it fixes #27024 and https://github.com/joomla/joomla-cms/pull/26077#issuecomment-544872087  also

Commit:
Merge branch '4.0-dev' into asset-4

Quy commented on build/build-modules-js/settings.json:
```diff
@@ -395,7 +508,7 @@
           "qrcode_UTF8.js": "js/qrcode_UTF8.js"
         },
-        "provideAssets": [
+        "provideAssets1": [
```
Is it correct with `1`?

Fedik replied:
nope, I will check

Quy commented on build/build-modules-js/settings.json:
```diff
@@ -104,13 +134,26 @@
           "public/assets/styles/choices.min.css": "css/choices.min.css"
         },
-        "provideAssets": [
+        "provideAssets":[
```
Some instances there is a space after the colon. Which is correct?

Fedik replied:
with space

Quy commented on libraries/src/Event/WebAsset/WebAssetRegistryAssetChanged.php:
```diff
@@ -0,0 +1,111 @@
+
+namespace Joomla\CMS\Event\WebAsset;
+
+defined('JPATH_PLATFORM') or die;
```
```suggestion
\defined('JPATH_PLATFORM') or die;
```

Quy commented on libraries/src/WebAsset/AssetItem/FormValidateAssetItem.php:
```diff
@@ -0,0 +1,43 @@
+
+namespace Joomla\CMS\WebAsset\AssetItem;
+
+defined('JPATH_PLATFORM') or die;
```
```suggestion
\defined('JPATH_PLATFORM') or die;
```

Fedik committed:
Update libraries/src/Event/WebAsset/WebAssetRegistryAssetChanged.php

Co-Authored-By: Quy <quy@fluxbb.org>

Fedik committed:
Update libraries/src/WebAsset/AssetItem/FormValidateAssetItem.php

Co-Authored-By: Quy <quy@fluxbb.org>

Fedik committed:
sync. Merge branch '4.0-dev' into asset-4

Fedik committed:
settings.json correction

Fedik committed:
Add registerAndUse<Type> magick

Fedik committed:
couple fixes

Fedik committed:
Add dummy asset to allow to extensions to use it as dependency to active template

Fedik commented:
**New feature**: an extension now can set `template.active` as dependency to current template, so their files will be loaded after template style/script files, even if extensions asset was enabled before template render.

SharkyKZ commented:
Getting this in frontend:

>Error: Fail parsing of asset registry file "media\vendor\joomla.asset.json". Property "type" are required for asset "awesomplete": Fail parsing of asset registry file "media\vendor\joomla.asset.json". Property "type" are required for asset "awesomplete"

SharkyKZ commented on administrator/templates/atum/joomla.asset.json:
```diff
@@ -6,40 +6,75 @@
+      "name": "template.active",
+      "type": "script",
+      "uri": "",
+      "description": "A dummy asset to allow to extensions to use it as dependency to active template",
```
```suggestion
      "description": "A dummy asset to allow extensions to use it as a dependency to active template",
```

brianteeman replied:
....as a dependency ...

SharkyKZ replied:
Right, thanks.

SharkyKZ commented on libraries/src/WebAsset/WebAssetRegistry.php:
```diff
@@ -265,22 +329,71 @@ protected function parseRegistryFile($path)
 			{
-				throw new \RuntimeException('Asset data file "' . $path . '" contains incorrect asset definition');
+				throw new \RuntimeException(
+					sprintf('Fail parsing of asset registry file "%s". Property "name" are required for asset index "%s"', $path, $i)
```
```suggestion
					sprintf('Failed parsing asset registry file "%s". Property "name" is required for asset index "%s"', $path, $i)
```

SharkyKZ commented on libraries/src/WebAsset/WebAssetRegistry.php:
```diff
@@ -265,22 +329,71 @@ protected function parseRegistryFile($path)
+			if (empty($item['type']))
+			{
+				throw new \RuntimeException(
+					sprintf('Fail parsing of asset registry file "%s". Property "type" are required for asset "%s"', $path, $item['name'])
```
```suggestion
					sprintf('Failed parsing asset registry file "%s". Property "type" is required for asset "%s"', $path, $item['name'])
```

SharkyKZ commented on libraries/src/WebAsset/WebAssetRegistry.php:
```diff
@@ -265,22 +329,71 @@ protected function parseRegistryFile($path)
 	}
+
+	/**
+	 * Dispatch an even to notify a listeners about an asset changes: new, remove, override
```
```suggestion
	 * Dispatch an event to notify listeners about asset changes: new, remove, override
```

Fedik committed:
Update administrator/templates/atum/joomla.asset.json

Co-Authored-By: SharkyKZ <sharkykz@gmail.com>

Fedik committed:
Update libraries/src/WebAsset/WebAssetRegistry.php

Co-Authored-By: SharkyKZ <sharkykz@gmail.com>

Fedik committed:
Update libraries/src/WebAsset/WebAssetRegistry.php

Co-Authored-By: SharkyKZ <sharkykz@gmail.com>

Fedik committed:
Update libraries/src/WebAsset/WebAssetRegistry.php

Co-Authored-By: SharkyKZ <sharkykz@gmail.com>

SharkyKZ commented:
With PR `jquery-noconflict` is not loading.

Fedik commented:
It does not required, but you can still load it if you need it

infograf768 committed:
Merge branch '4.0-dev' into asset-4

jwaisner commented:
I have tested this item :white_check_mark: successfully on 9de1195a3d2ddadbf2ff54568b86c8eca5352c47<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25775">issues.joomla.org/tracker/joomla-cms/25775</a>.</sub>

wilsonge commented:
I have to say with this I'm slightly concerned that we are starting to make things much more complicated in the JSON structure. But at the same time as we're solving concrete problems I'm not sure there's much we can do about it. Can we please get the documentation and the JSON Schema updated please for this new approach

Thankyou for taking the time to investigating michael's review. It's very much appreciated by me :)

Fedik commented:
Schema update are there https://github.com/joomla/schemas/pull/7
But do not merge for now please. I have to update a bit, about webcomponents ( to cover #27268)

>  I'm slightly concerned that we are starting to make things much more complicated in the JSON structure

yeah, me to, a bit :)
But it not that much complicated as much more content it contain now, because now an asset can contain only one file.
Also, because of this, I have introduced a "preset", a kind of collection that can "hold" a multiple different asset. It more like a "helper".
will see how it will work

the doc page I will look

roland-d commented:
@Fedik Did you write up the documentation for the JSON format?

Fedik commented:
Documentation I have updated, it is there https://docs.joomla.org/J4.x:Web_Assets
And JSON schema are there https://github.com/joomla/schemas/blob/master/json-schema/web_assets.json 

In documentation there just some examples of JSON, not sure what to write about schema.

roland-d commented:
@Fedik Thank you, I could not find the article on the docs.

Fedik commented:
btw, here is the "public link" to schema https://developer.joomla.org/schemas/json-schema/web_assets.json

roland-d commented:
@Fedik The only thing that is unclear to me is this:
```
"dependencies": [
        "core",
		"bootstrap"
      ],
```
How do I know which dependencies are available? The bootstrap is giving a fatal error now. I have no idea where to check what it is looking for. This might be something we can explain in the docs as well.

Fedik commented:
It some how this section https://docs.joomla.org/J4.x:Web_Assets#Register_an_asset
bootstrap actually part of vendors json `media/vendor/joomla.asset.json`

It should be available on `on first access to WebAssetRegistry`

But you have to check that file to get a correct name, there a couple of `bootstrap` variations, eg: `bootstrap.css` (styles), `bootstrap.css.grid` (another style), `bootstrap.js` (script), `bootstrap.js.bundle` (another script)

I had an idea to extend a debug plugin to print out all available assets, but this for some time later maybe, maybe for 4.2 or so :smiley:

Fedik commented:
and the `core` asset is part of system json `media/system/joomla.asset.json`

roland-d commented:
Hey @Fedik Thank you for the feedback so far. 
> It should be available on on first access to WebAssetRegistry

That doesn't seem to be the case. When I inspect `$wa` in the index.php of the template, it has some json files loaded but nothing parsed.
![image](https://user-images.githubusercontent.com/359377/73764681-b7484780-4773-11ea-970e-8ac0f4e5b22c.png)

When I add `core` as dependency in the `templates/cassiopeia/joomla.asset.json` file like this:
```
"name": "template.cassiopeia.ltr",
      "type": "preset",
      "dependencies": [
		"core",
        "template.cassiopeia.ltr#style",
        "template.cassiopeia#script"
      ]
```
It also bails with an error 
```
Error: Unsatisfied dependency "core" for an asset "template.cassiopeia.ltr" of type "preset": Unsatisfied dependency "core" for an asset "template.cassiopeia.ltr" of type "preset"
```
It looks like, the json files need to be parsed first before I can load any preset/asset with a dependency. Any ideas about this?

Fedik commented:
> It should be available on on first access to WebAssetRegistry
> When I inspect $wa in the index.php of the template, it has some json files loaded but nothing parsed.

that correct, they will be parsed on "first access", mean on first `$registry->get()` call, that will be performed on any `$wa->useScript('foobar')` call

> When I add core as dependency ... It also fails with an error.

That also correct, see what type of asset do you use. If your asset is a `script` then all dependency also must be `script`, if your asset are `style` then all dependency also must be type of `style` and so on, `preset` for presets.
There no support of "cross type" dependency (except for presets).

Your example:
```
"name": "template.cassiopeia.ltr",
"type": "preset", <----- Asset type
"dependencies": [
  "core",   <------ Looking for another "preset" with name "core", not for "script" asset
  "template.cassiopeia.ltr#style",
  "template.cassiopeia#script"
]
```
As error say, there no "preset" with name "core".
Here is about presets https://docs.joomla.org/J4.x:Web_Assets#Working_with_a_presets with explanation what is special in presets.

Correct would be: 
```
"name": "template.cassiopeia.ltr",
"type": "preset",
"dependencies": [
  "core#script",
  "template.cassiopeia.ltr#style",
  "template.cassiopeia#script"
]
```
But again, this only for "preset" type, other asset types does not support "cross type" dependency.

roland-d commented:
@Fedik The exteneded explanation is much appreciated. This gives me a lot of insight and it is working now.

Fedik commented:
It was a bit different in first implementation,
In first implementation an asset hold diferent files css, js, and then after some discussions I have change to handle `1 asset` =  `1 file`
